### PR TITLE
Add downsample to ILMs

### PR DESCRIPTION
**Pull request by xaviervolvo**
There is a parameter for the hot, warm and cold phases called downsample.

According to the description:

"Roll up documents within a fixed interval to a single summary document. Reduces the index footprint by storing time series data at reduced granularity."

As of 0.11.0 there is no support for this parameter in the proivider, this PR adds it.

### CODE REVIEW

**Comment by xaviervolvo** [comment]
Hello, I signed the agreement but this seems stuck. Let me know if i need to do anything else.

File: internal/elasticsearch/index/ilm.go [hunk 1]
```diff
@@ -389,6 +389,26 @@ var supportedActions = map[string]*schema.Schema{
 			},
 		},
 	},
+	"downsample": {
+		Description: "Roll up documents within a fixed interval to a single summary document. Reduces the index footprint by storing time series data at reduced granularity.",
+		Type:        schema.TypeList,
+		Optional:    true,
+		MaxItems:    2,
```
Comment (tobio) [inline internal/elasticsearch/index/ilm.go]: Why 2? IIUC there's no case where a user should specify more than a single `downsample` block right?

File: internal/elasticsearch/index/ilm_test.go [hunk 2]
```diff
@@ -128,6 +130,11 @@ resource "elasticstack_elasticsearch_index_lifecycle" "test" {
       max_age = "1d"
     }
 
+		downsample {
```
Comment (tobio) [inline internal/elasticsearch/index/ilm_test.go]: nit: Can we fix up the indentation on this block please.

**Review by tobio — COMMENTED** [review]
Thanks for adding this, small question on the schema.

**Comment by tobio** [comment]
@xaviervolvo are you able to add an entry to the CHANGELOG.md file for this one too?

**Comment by xaviervolvo** [comment]
I think I addressed all the questions, thank you.

**Review by tobio — APPROVED** [review]
LGTM

**Review by tobio — DISMISSED** [review]
Still LGTM, just was regenerated docs.

**Comment by xaviervolvo** [comment]
I see that acceptance tests are failing, is there anything I can do to solve that?

**Comment by martijnbakkerelastic** [comment]
@tobio When briefly discussing the current status of the PR with Volvo Cars, they shared the feedback on the failed checks that this might have to do with downsample not being implemented in version prior to 8.10. Could this maybe be the cause and if so how can we move this PR forward?

**Comment by tobio** [comment]
> they shared the feedback on the failed checks that this might have to do with downsample not being implemented in version prior to 8.10. Could this maybe be the cause and if so how can we move this PR forward?

It looks like there's three cases here:

* [8.4 and lower](https://github.com/elastic/terraform-provider-elasticstack/actions/runs/7719631590/job/21043205348?pr=538): No support for downsample.
* [8.5 - 8.9](https://github.com/elastic/terraform-provider-elasticstack/actions/runs/7719631590/job/21043206307?pr=538): No support for the `wait_timeout` attribute
* 8.10 and higher: Full support

We don't usually restrict provider functionality based on the target cluster version (i.e there's no provider level version checks, the resource will just fail). That's not ideal, nicer errors would of course be nicer, I'm just stating the status quo. If a nice error message was added here then that would be lovely, but not necessary. 

Instead, we usually filter the acceptance tests, the ILM test already does this for the [shards per node](https://github.com/elastic/terraform-provider-elasticstack/blob/main/internal/elasticsearch/index/ilm_test.go#L65) case. It looks like we should do the same here, i.e create a new test (or two given the 8.5-8.9 case above) case explicitly validating the downscale attributes work, and skipping them on unsupported versions.

**Comment by xaviervolvo** [comment]
I understand. I haven't familiarised myself with the test setup to the point to see how to filter out tests depending on version?  Can you take care of it or would you give me some pointers? Thank you.

**Comment by tobio** [comment]
> Can you take care of it or would you give me some pointers

If you have some time to wrap this one up then it'll get merged and released faster :) 

[This](https://github.com/elastic/terraform-provider-elasticstack/blob/main/internal/elasticsearch/index/ilm_test.go#L65) test case shows how we're skipping test cases based on the target stack version. Let me know if you've got any questions for that, but we've just been defining a minimum version and using this skip mechanism elsewhere.

**Comment by xaviervolvo** [comment]
I created skip versions functions and added the test cases for the different versions as mentioned in the comment above.

**Review by tobio — APPROVED** [review]
LGTM

**Comment by tobio** [comment]
@xaviervolvo thanks for making the time for this addition!

**Pull request merged**